File: The Core/Memory.cpp
#include "Memory.h"
#include "Converter.h"

const int SIZE_OF_MEMORY(65535);

Memory::Memory(){
	this->clear();
}

Memory::Memory(Memory &copyingMemory){
	memory = copyingMemory.get_memory();
	names = copyingMemory.get_names();
}

void Memory::clear(){
	for (int i = 0; i < SIZE_OF_MEMORY; i++)
		this->memory[i] = nullptr;
}

MemoryItem* Memory::get(int position){
	this->check(position);
	
	if (this->memory[position].get() == nullptr)
		throw std::out_of_range("Appealing to a non-existent adress");

	return this->memory[position].get();
}

std::unordered_map<int, std::shared_ptr<MemoryItem> > Memory::get_memory(){
	return memory;
}

std::unordered_map<int, std::string> Memory::get_names(){
	return names;
}

void Memory::set(int position, MemoryItem* item){
	this->check(position);

	this->memory[position] = std::shared_ptr<MemoryItem>(item);
}

bool Memory::is_memory_cell_empty(int position){
	this->check(position);
	return this->memory[position].get() == nullptr;
}

void Memory::set_name(int position, std::string name){
	this->check(position);

	this->names[position] = name;
}

std::string Memory::get_name(int position)const{
	this->check(position);
	
	if (this->names.find(position) == this->names.end())
		return Converter::to_hex(position, 4);
	else
		return this->names.find(position)->second;
}

void Memory::check(int position)const{
	if (position < 0 || position > SIZE_OF_MEMORY)
		throw std::exception("Memory bounds violation");
}

File: Windows - GUI CS/Memory/Memory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LM_GUI_UP
{
    class Memory
    {
        const int SIZE_OF_MEMORY = 65536;

        private MemoryItem[] _memory;
        private Dictionary<int, string> _names;

        public Memory()
        {
            _memory = new MemoryItem[SIZE_OF_MEMORY];
            _names = new Dictionary<int, string>();

            ClearMemory();
        }

        #region Methods for memory
        public MemoryItem GetCell(int position)
        {
            Check(position);

            if (_memory[position] == null)
                throw new Exception("memory error: appealing to a non-existent adress.");

            return _memory[position];
        }
        public void SetCell(int position, MemoryItem item)
        {
            Check(position);

            _memory[position] = item;
        }
        public void ClearMemory()
        {
            for (int i = 0; i < SIZE_OF_MEMORY; i++)
                _memory[i] = null;

            _names.Clear();
        }
        public bool IsMemoryCellEmpty(int position)
        {
            Check(position);

            return _memory[position] == null;
        }
        #endregion

        #region Methods for names
        public void SetName(int position, string name)
        {
            Check(position);

            if (_names.ContainsKey(position))
                _names[position] = name;
            else
                _names.Add(position, name);
        }

        public string GetName(int position)
        {
            Check(position);

            if (_names.ContainsKey(position))
                return _names[position]; 
            else
                return TextFormat.addZeros(Convert.ToString(position, 16).ToUpper(),4);
        }

        #region information about variables

        private string StringInDecNumber(string address, string name, int value)
        {
            return string.Format("{0} ( {1} ) := {2}", address, name, value);
        }

        private string StringInAdditionalCode(string address, string name, int value)
        {
            return string.Format("{0} ( {1} ) := {2} ({3})", address, name, value,TextFormat.additionalCode16(value));
        }

        private List<string> CreateVarInfo(Func<string, string, int, string> stringFormat)
        {
            List<string> varInfo = new List<string>();

            for (int i = 0; i < _memory.Length; i++)
                if (_memory[i] != null && _memory[i] is Variable)
                    varInfo.Add(stringFormat(TextFormat.addZeros(Convert.ToString(i, 16).ToUpper(), 4),GetName(i), _memory[i].GetValue()));

            return varInfo;
        }

        public List<string> GetVarInfo()
        {
            return CreateVarInfo(StringInDecNumber);
        }

        public List<string> GetVarInfoInAdditionalCode()
        {
            return CreateVarInfo(StringInAdditionalCode);
        }

        #endregion

        public List<string> GetCmdInfo()
        {
            List<string> cmdInfo = new List<string>();

            for (int i = 0; i < _memory.Length; i++)
                if (_memory[i] != null && _memory[i] is Command)
                    cmdInfo.Add(string.Format("{0} := {1}", TextFormat.addZeros(Convert.ToString(i, 16).ToUpper(), 4), _memory[i].ToString()));

            return cmdInfo;
        }

        #endregion
        public void Check(int position)
        {
            if (position < 0 || position > SIZE_OF_MEMORY)
                throw new Exception("Memory error: Memory bounds violation.");
        }
    }
}
